File: documentation/example-module/example-module.cpp
/*
 *   Copyright (C) 2022 Event-driven Perception for Robotics
 *   Author: arren.glover@iit.it
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU Lesser General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU Lesser General Public License
 *   along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <yarp/os/all.h>
#include <event-driven/core.h>
#include <thread>
#include <mutex>
#include <condition_variable>

using namespace ev;
using namespace yarp::os;

class exampleModule : public RFModule {

private:

    //yarp ports
    ev::window<AE> input_port;
    ev::BufferedPort<AE> output_port;

    //threads
    std::mutex m;
    std::condition_variable signal;
    std::thread event_thread;
    std::thread proc_thread;
    double current_ts{0.0}, previous_ts{0.0};

    //internal data
    cv::Mat my_eros;
    bool example_flag;
    double example_parameter;

public:

    exampleModule() {}

    virtual bool configure(yarp::os::ResourceFinder& rf)
    {
        if(rf.check("h") || rf.check("help")) 
        {
            yInfo() << "help message";
            yInfo() << "variable 1 <int> : description";
        }

        /* initialize yarp network */
        if (!yarp::os::Network::checkNetwork(2.0)) {
            std::cout << "Could not connect to YARP" << std::endl;
            return false;
        }

        //set the module name used to name ports
        setName((rf.check("name", Value("/example-module")).asString()).c_str());

        //open io ports
        if(!input_port.open(getName() + "/AE:i")) {
            yError() << "Could not open input port";
            return false;
        }
        output_port.setWriteType(AE::tag);
        if(!output_port.open(getName() + "/AE:o")) {
            yError() << "Could not open input port";
            return false;
        }

        //read flags and parameters
        example_flag = rf.check("example_flag") &&
                rf.check("example_flag", Value(true)).asBool();
        double default_value = 0.1;
        example_parameter = rf.check("example_parameter",
                                     Value(default_value)).asFloat64();

        //do any other set-up required here
        cv::namedWindow(getName(), cv::WINDOW_NORMAL);
        cv::resizeWindow(getName(), cv::Size(640, 480));
        event_thread = std::thread([this]event_process);
        proc_thread = std::thread([this]secondary_process);

        //start the asynchronous and synchronous threads
        return Thread::start();
    }

    virtual double getPeriod()
    {
        return 1.0; //period of synchrnous thread
    }

    bool interruptModule()
    {
        //if the module is asked to stop ask the asynchrnous thread to stop
        input_port.close();
        event_thread.join();
        proc_thread.join();
        return Thread::stop();
    }

    //synchronous thread (threaded) critical function
    virtual bool updateModule()
    {
        cv::imshow(getName(), my_eros);
        cv::waitKey(1);

        //add any synchronous operations here, visualisation, debug out prints
        return Thread::isRunning();
    }

    //critical (threaded) function
    void event_process()
    {
        while(!isStopping()) 
        {
            ev::info stats = input_port.readAll(true);

            std::unique_lock<std::mutex> lk(m);
            for(auto &v : input_port)
                ev::eros.update();
            current_ts = stats.timestamp;
            lk.unlock();
            signal.notify_one();
        }
    }

    //critical (threaded) function
    void secondary_process() 
    {
        while(!isStopping()) 
        {
            //get the data structure
            std::unique_lock<std::mutex> lk(m);
            signal.wait(lk, [this](return current_ts > previous_ts || isStopping()););
            if(isStopping()) break;
            previous_ts = current_ts;
            ev::eros.getSurface().copyTo(my_eros);
            lk.unlock();

            //further processing

            //event-driven output if necessary
            ev::packet<AE> &output_packet = output_port.prepare();
            output_packet.clear();
            AE ae;
            output_packet.duration();
            output_packet.timestamp();
            output_packet.push_back(ae);
            output_port.write();
        }
    }
};

int main(int argc, char * argv[])
{
    /* prepare and configure the resource finder */
    yarp::os::ResourceFinder rf;
    rf.configure( argc, argv );

    /* create the module */
    exampleModule instance;
    return instance.runModule(rf);
}

File: documentation/example-module-py/example-module.py
import yarp
import sys
import numpy as np
import matplotlib.pyplot as plt
import threading


class ExampleModule(yarp.RFModule, threading.Thread):

    def __init__(self):
        yarp.RFModule.__init__(self)
        threading.Thread.__init__(self)
        self.image = np.zeros((240, 304))
        self.image_buf = np.zeros((240, 304))
        self.im = plt.imshow(self.image, cmap='gist_gray_r', vmin=0, vmax=1)
        self.input_port = yarp.BufferedPortBottle()
        self.rpc_port = yarp.RpcServer()
        self.mutex = threading.Lock()

    def configure(self, rf):
        # set the module name used to name ports
        self.setName((rf.check("name", yarp.Value("/exampleModule")).asString()))

        # open io ports
        if not self.input_port.open(self.getName() + "/AE:i"):
            print("Could not open input port")
            return False
        self.input_port.setStrict()

        if not self.rpc_port.open(self.getName() + "/rpc"):
            print("Could not open rpc port")
            return False
        self.attach_rpc_server(self.rpc_port)  # rpc port receives command in the respond method

        # read flags and parameters
        example_flag = rf.check("example_flag") and rf.check("example_flag", yarp.Value(True)).asBool()
        default_value = 0.1
        example_parameter = rf.check("example_parameter", yarp.Value(default_value)).asDouble()

        # do any other set-up required here
        # start the asynchronous and synchronous threads
        self.start()
        return True

    def respond(self, command, reply):
        # Add any command you want to receive from rpc here
        print(command.toString())
        reply.addString('ok')
        return True

    def getPeriod(self):
        return 0.03  # period of synchronous thread, return 0 update module called as fast as it can

    def interruptModule(self):
        # interrupting all the ports
        self.input_port.interrupt()
        self.rpc_port.interrupt()
        return True

    def close(self):
        # closing ports
        self.input_port.close()
        self.rpc_port.close()
        return True

    def updateModule(self):
        # synchronous update called every get period seconds.

        # Put visualization, debug prints, etc... here
        self.mutex.acquire()
        self.im.set_data(self.image)
        self.mutex.release()
        plt.draw()
        plt.pause(1e-3)
        return True

    def run(self):
        # asynchronous thread runs as fast as it can

        while self.is_alive():
            reads = max(1, self.input_port.getPendingReads())  # Make sure that we try to read at least one packet

            if reads > 10:
                print("WARNING ", reads)

            self.image_buf.fill(0.5)

            for i in range(reads):
                bot = self.input_port.read()
                # Data in the bottle is organized as <event_type> (<timestamp 1> <event 1> .... <timestamp n> <event n>)
                vType = bot.get(0).asString()
                event_bottle = np.array(bot.get(1).toString().split(' ')).astype(int).reshape(-1, 2)
                timestamps = event_bottle[:, 0]
                events = event_bottle[:, 1]
                x = events >> 12 & 0xFF
                y = events >> 1 & 0x1FF
                pol = events & 0x01
                self.image_buf[x, y] = pol

            self.mutex.acquire()
            self.image = self.image_buf.copy()  # self.image is a shared resource between threads
            self.mutex.release()


if __name__ == '__main__':
    # Initialise YARP
    yarp.Network.init()
    if not yarp.Network.checkNetwork(2):
        print("Could not find network! Run yarpserver and try again.")
        exit(-1)

    # prepare and configure the resource finder
    rf = yarp.ResourceFinder()
    rf.setVerbose(False)
    rf.setDefaultContext("eventdriven")
    rf.setDefaultConfigFile("exampleModule.ini")
    rf.configure(sys.argv)

    # create the module
    module = ExampleModule()
    module.runModule(rf)
